tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,361 @@
+    """Extract and merge patches using fixed sized windows for images and labels.
+
+    Args:
+        stride_h(int): stride in horizontal direction for patch extraction.
+        stride_w(int): stride in vertical direction for patch extraction.
+    """
+
+    def __init__(
+        self,
+        input_image,
+        img_patch_h,
+        img_patch_w,
+        pad_y=0,
+        pad_x=0,
+        stride_h=1,
+        stride_w=1,
+    ):
+        super().__init__(
+            input_image=input_image,
+            img_patch_h=img_patch_h,
+            img_patch_w=img_patch_w,
+            pad_y=pad_y,
+            pad_x=pad_x,
+        )
+        self.stride_h = stride_h
+        self.stride_w = stride_w
+
+    # def extract_patches(self, input_img, labels=None):
+    #     raise NotImplementedError
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class VariableWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using variable sized windows for images and labels.
+
+    Args:
+        stride_h: stride in horizontal direction for patch extraction
+        stride_w: stride in vertical direction for patch extraction
+        label_patch_h: network output label height
+        label_patch_w: network output label width
+    """
+
+    def __init__(
+        self,
+        input_image,
+        img_patch_h,
+        img_patch_w,
+        pad_y=0,
+        pad_x=0,
+        stride_h=1,
+        stride_w=1,
+        label_patch_h=None,
+        label_patch_w=None,
+    ):
+        super().__init__(
+            input_image=input_image,
+            img_patch_h=img_patch_h,
+            img_patch_w=img_patch_w,
+            pad_y=pad_y,
+            pad_x=pad_x,
+        )
+        self.stride_h = stride_h
+        self.stride_w = stride_w
+        self.label_patch_h = label_patch_h
+        self.label_patch_w = label_patch_w
+
+    # def extract_patches(self, input_img, labels=None):
+    #     raise NotImplementedError
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class PointsPatchExtractor(PatchExtractor):
+    """Extracting patches with specified points as a centre.
+
+    Args:
+        labels(ndarray, pd.DataFrame, str, pathlib.Path): contains location and/or
+         type of patch. Input can be path to a csv or json file.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification, default=9 (centre of patch and all the eight neighbours as
+         centre).
+
+    Attributes:
+        labels(pd.DataFrame): A table containing location and/or type of patch.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification.
+
+    """
+
+    def __init__(
+        self,
+        input_image,
+        labels,
+        img_patch_h=224,
```

Comment:
Convert to tuple

Reply:
Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/9f1865f8bf2eb6b49eb5b19b485e3f522a524b31